### PR TITLE
provide party and package management in KVUtils and participant-state v1

### PR DESCRIPTION
**Pull request by mziolekda**
provide synchronous interfaces for party and package management on the participant-state
add KVUtils support for party and package management

### CODE REVIEW

File: ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/KeyValueCommitting.scala [hunk 1]
```diff
@@ -146,6 +142,157 @@ object KeyValueCommitting {
...
+      inputState: Map[DamlStateKey, Option[DamlStateValue]]
+  ): (DamlLogEntry, Map[DamlStateKey, DamlStateValue]) = {
+    val submissionId = packageUploadEntry.getSubmissionId
+    def tracelog(msg: String): Unit =
+      logger.trace(
+        s"processPackageUpload[entryId=${prettyEntryId(entryId)}, submId=$submissionId]: $msg")
+
+    val archives = packageUploadEntry.getArchivesList.asScala
+    val packagesValidityResult =
+      archives.foldLeft[(Boolean, String)]((true, ""))(
```
Comment (dajmaki) [inline ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/KeyValueCommitting.scala]: could as well just do `foldLeft("")` and `packagesValidityError.isEmpty` etc.

File: ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/KeyValueCommitting.scala [hunk 2]
```diff
@@ -146,6 +142,157 @@ object KeyValueCommitting {
...
+        (acc, archive) =>
+          if (archive.getPayload.isEmpty) (false, acc._2 ++ s"empty package '${archive.getHash}';")
+          else acc)
+
+    if (packagesValidityResult._1) {
```
Comment (dajmaki) [inline ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/KeyValueCommitting.scala]: swap this so we can read it more naturally: check for errors, handle the error, otherwise continue.

File: ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/KeyValueCommitting.scala [hunk 3]
```diff
@@ -146,6 +142,157 @@ object KeyValueCommitting {
...
+      (
+        DamlLogEntry.newBuilder
+          .setRecordTime(buildTimestamp(recordTime))
+          .setPackageUploadEntry(
+            DamlPackageUploadEntry.newBuilder
+              .setSubmissionId(submissionId)
+              .addAllArchives(filteredArchives.asJava)
+              .setSourceDescription(packageUploadEntry.getSourceDescription)
+              .setParticipantId(packageUploadEntry.getParticipantId)
+              .build
```
Comment (dajmaki) [inline ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/KeyValueCommitting.scala]: tip: inner .build not needed as `setFoo` takes a `FooOrBuilder`

File: ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/KeyValueCommitting.scala [hunk 4]
```diff
@@ -146,6 +142,157 @@ object KeyValueCommitting {
...
+          .map(
+            archive =>
+              (
+                DamlStateKey.newBuilder.setPackageId(archive.getHash).build,
+                DamlStateValue.newBuilder.setArchive(archive).build
+            )
+          )(breakOut)
+      )
+    } else {
+      tracelog(s"Packages upload failed, invalid package submitted")
```
Comment (dajmaki) [inline ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/KeyValueCommitting.scala]: log what invalid package it was and why it was invalid.

File: ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/KeyValueCommitting.scala [hunk 5]
```diff
@@ -146,6 +142,157 @@ object KeyValueCommitting {
...
+      buildPackageRejectionLogEntry(recordTime, packageUploadEntry, {
+        _.setInvalidPackage(packagesValidityResult._2)
+      })
```
Comment (dajmaki) [inline ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/KeyValueCommitting.scala]: are the curlybraces needed?

File: ledger/participant-state/src/main/scala/com/daml/ledger/participant/state/v1/RejectionReason.scala [hunk 6]
```diff
@@ -7,7 +7,7 @@ package com.daml.ledger.participant.state.v1
   *
   * Used to provide details for [[Update.CommandRejected]].
   */
-sealed trait RejectionReason extends Product with Serializable {
+sealed abstract class RejectionReason extends Product with Serializable {
```
Comment (dajmaki) [inline ledger/participant-state/src/main/scala/com/daml/ledger/participant/state/v1/RejectionReason.scala]: why the change?

File: ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/InMemoryKVParticipantState.scala [hunk 7]
```diff
@@ -125,6 +139,62 @@ class InMemoryKVParticipantState(implicit system: ActorSystem, mat: Materializer
       )
     )
 
+  /** Akka actor that matches the requests for party allocation
+    * with asynchronous responses delivered within the log entries.
+    */
+  class ResponseMatcher extends Actor {
```
Comment (dajmaki) [inline ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/InMemoryKVParticipantState.scala]: I'd be interested in seeing how the Sawtooth implementation of this logic ends up looking like and whether we can abstract it out and use it for other implementations as well.

File: ledger/participant-state/kvutils/src/main/protobuf/daml_kvutils.proto [hunk 8]
```diff
@@ -208,3 +290,10 @@ message DamlContractState {
 
 }
 
+// Stored information about a given party.
+message DamlPartyValue {
```
Comment (dajmaki) [inline ledger/participant-state/kvutils/src/main/protobuf/daml_kvutils.proto]: I'd prefer a more descriptive name.. like DamlPartyAllocation? Not DamlPartyState as this is immutable.

File: ledger/participant-state/src/main/scala/com/daml/ledger/participant/state/v1/RejectionReason.scala (hunk 6 above)
Comment (mziolekda) [reply ledger/participant-state/src/main/scala/com/daml/ledger/participant/state/v1/RejectionReason.scala]: it is the only one of many siblings that is a trait; abstract class is preferred due to performance on JVM

File: ledger/participant-state/kvutils/src/main/protobuf/daml_kvutils.proto [hunk 9]
```diff
@@ -121,12 +130,45 @@ message DamlRejectionEntry {
...
 }
 
-// A public package upload. We don't store dars as they would
-// not allow for easy deduplication of DAML-LF packages.
+// A public package upload.
+// DAML-LF packages will be deduplicated based on their hashes. Only unique
+// ones will be added to the KV storage. Deduplication has no impact on the
+// upload result.
 message DamlPackageUploadEntry {
-  repeated daml_lf.Archive archives = 1;
```
Comment (gerolf-da) [inline ledger/participant-state/kvutils/src/main/protobuf/daml_kvutils.proto]: Same as above. Do not re-purpose field ids.

File: ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/KeyValueCommitting.scala [hunk 10]
```diff
@@ -146,6 +142,157 @@ object KeyValueCommitting {
...
+    val submissionId = packageUploadEntry.getSubmissionId
+    def tracelog(msg: String): Unit =
+      logger.trace(
+        s"processPackageUpload[entryId=${prettyEntryId(entryId)}, submId=$submissionId]: $msg")
+
+    val archives = packageUploadEntry.getArchivesList.asScala
+    val packagesValidityResult =
+      archives.foldLeft[(Boolean, String)]((true, ""))(
+        (acc, archive) =>
+          if (archive.getPayload.isEmpty) (false, acc._2 ++ s"empty package '${archive.getHash}';")
```
Comment (rautenrieth-da) [inline ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/KeyValueCommitting.scala]: Is that enough to check the validity of the payload? Shouldn't you try to parse the archive?

File: ledger/participant-state/kvutils/src/main/protobuf/daml_kvutils.proto [hunk 11]
```diff
@@ -58,11 +59,19 @@ message DamlLogEntry {
...
 
     // A collection of DAML-LF archives uploaded together by a participant.
-    // Can be converted into `Update.PublicPackageUploaded`.
     DamlPackageUploadEntry package_upload_entry = 4;
 
+    // A rejection of package upload
+    DamlPackageUploadRejectionEntry package_upload_rejection_entry = 5;
+
     // A DAML configuration change.
-    DamlConfigurationEntry configuration_entry = 5;
```
[the remark this answers is not in the document]
Comment (mziolekda) [reply ledger/participant-state/kvutils/src/main/protobuf/daml_kvutils.proto]: Yes, the interface is deemed unstable and we control all consumers

File: ledger/participant-state/kvutils/src/main/protobuf/daml_kvutils.proto [hunk 12]
```diff
@@ -58,11 +59,19 @@ message DamlLogEntry {
...
+    DamlConfigurationEntry configuration_entry = 6;
```
[the remark this answers is not in the document]
Comment (mziolekda) [reply ledger/participant-state/kvutils/src/main/protobuf/daml_kvutils.proto]: We control all consumers, interface is considered unstable at the moment

File: ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/KeyValueCommitting.scala (hunk 10 above)
Comment (mziolekda) [reply ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/KeyValueCommitting.scala]: Well, the proper validation would go through a transitive closure of all the packages. It will have to be added at a later stage. We have agreed with Jussi to do the minimum at the moment. I can add a clear TODO in the code

File: ledger/participant-state/kvutils/src/main/protobuf/daml_kvutils.proto [hunk 13]
```diff
@@ -121,12 +130,45 @@ message DamlRejectionEntry {
...
+  // An id selected by the submitter, can be used for matching rejection with
+  // the original upload entry.
+  string submission_id = 1;
+
+  // Uploading participant's id.
+  string participant_id = 2;
+
+  oneof reason {
+    // One of the uploaded packages is not valid.
+    string invalid_package = 3;
```
[the remark this answers is not in the document]
Comment (mziolekda) [reply ledger/participant-state/kvutils/src/main/protobuf/daml_kvutils.proto]: What should happen with the preexisting transaction rejection entries. Should they also change. For consistency sake I used the same pattern as in:
`message DamlRejectionEntry {
  DamlSubmitterInfo submitter_info = 1;

  oneof reason {
    // NOTE(JM): Not all payloads are used, but marked as 'string' for
    // extensibility.
    string inconsistent = 2;
    string disputed = 3;
    string resources_exhausted = 4;
    string maximum_record_time_exceeded = 5;
    string duplicate_command = 6;
    string party_not_known_on_ledger = 7;
    string submitter_cannot_act_via_participant = 8;
  }
}`

File: ledger/participant-state/kvutils/src/main/protobuf/daml_kvutils.proto [hunk 14]
```diff
@@ -208,3 +290,10 @@ message DamlContractState {
 
 }
 
+// Stored information about a given party.
+message DamlPartyValue {
+  // Id of participant where the party is hosted.
+  string participant_id = 1;
+  // A display name associated with the given party.
+  string display_name = 2;
```
[the remark this answers is not in the document]
Comment (mziolekda) [reply ledger/participant-state/kvutils/src/main/protobuf/daml_kvutils.proto]: It is not that simple as it would seem. Modification of the party record has to be verifiable by the all the committers/validators. For that you need a notion of LET for party allocation. I will a comment about it without addressing it ATM.

File: ledger/participant-state/kvutils/src/main/protobuf/daml_kvutils.proto (hunk 13 above)
[the remark this answers is not in the document]
Comment (mziolekda) [reply ledger/participant-state/kvutils/src/main/protobuf/daml_kvutils.proto]: Adding the same pattern to all the errors.

File: ledger/participant-state/kvutils/src/main/protobuf/daml_kvutils.proto [hunk 15]
```diff
@@ -34,6 +34,7 @@ message DamlSubmission {
     DamlTransactionEntry transaction_entry = 3;
     DamlPackageUploadEntry package_upload_entry = 4;
     DamlConfigurationEntry configuration_entry = 5;
+    DamlPartyAllocationEntry party_allocation_entry = 6;
```
[the remark this answers is not in the document]
Comment (mziolekda) [reply ledger/participant-state/kvutils/src/main/protobuf/daml_kvutils.proto]: Will be adding it in a follow-up PR

File: ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/KeyValueCommitting.scala (hunk 4 above)
Comment (mziolekda) [reply ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/KeyValueCommitting.scala]: Done

File: ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/KeyValueCommitting.scala (hunk 5 above)
Comment (mziolekda) [reply ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/KeyValueCommitting.scala]: Done

File: ledger/participant-state/kvutils/src/main/protobuf/daml_kvutils.proto (hunk 8 above)
Comment (mziolekda) [reply ledger/participant-state/kvutils/src/main/protobuf/daml_kvutils.proto]: OK
